### DISCUSSION TITLE
[@types/lodash] fp-types for propertyOf have wrong argument order

### DISCUSSION THREAD

**Discussion by primeapple** [DefinitelyTyped/DefinitelyTyped#62373] [Issues with a @types package]
Ok, so hear me out.
I was typing our codebase today and found weird type complaints of they LSP server.

The types for  `propertyOf` (doc: https://lodash.com/docs/4.17.15#propertyOf ) seem to have a wrong argument order.
Consider the following code, with lodash 4.17.21 installed
```javascript
import { propertyOf } from 'lodash/fp';
// correct usage of this method
propertyOf({"a": 1})("a")); // -> 1
```
The If I ask for the type of `propertyOf` here I get the following:
```
(alias) propertyOf<object, never>(path: [never]): _.LodashPropertyOf1x1<object, never> (+23 overloads)
```
As you can see it expects a `path` as first argument but get's an object. There is not a single type for `propertyOf` that has an object as first argument, they all have it as second argument.
You can see the types here: 

```typescript
    interface LodashPropertyOf {
        <TObject extends object, TKey extends keyof TObject>(path: TKey | [TKey]): LodashPropertyOf1x1<TObject, TKey>;
        <TObject extends object>(path: lodash.__, object: TObject): LodashPropertyOf1x2<TObject>;
        <TObject extends object, TKey extends keyof TObject>(path: TKey | [TKey], object: TObject): TObject[TKey];
        <TObject extends object>(path: lodash.__, object: TObject | null | undefined): LodashPropertyOf2x2<TObject>;
        <TObject extends object, TKey extends keyof TObject>(path: TKey | [TKey], object: TObject | null | undefined): TObject[TKey] | undefined;
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1]>(path: [TKey1, TKey2]): LodashPropertyOf3x1<TObject, TKey1, TKey2>;
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1]>(path: [TKey1, TKey2], object: TObject): TObject[TKey1][TKey2];
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1]>(path: [TKey1, TKey2], object: TObject | null | undefined): TObject[TKey1][TKey2] | undefined;
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1], TKey3 extends keyof TObject[TKey1][TKey2]>(path: [TKey1, TKey2, TKey3]): LodashPropertyOf5x1<TObject, TKey1, TKey2, TKey3>;
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1], TKey3 extends keyof TObject[TKey1][TKey2]>(path: [TKey1, TKey2, TKey3], object: TObject): TObject[TKey1][TKey2][TKey3];
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1], TKey3 extends keyof TObject[TKey1][TKey2]>(path: [TKey1, TKey2, TKey3], object: TObject | null | undefined): TObject[TKey1][TKey2][TKey3] | undefined;
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1], TKey3 extends keyof TObject[TKey1][TKey2], TKey4 extends keyof TObject[TKey1][TKey2][TKey3]>(path: [TKey1, TKey2, TKey3, TKey4]): LodashPropertyOf7x1<TObject, TKey1, TKey2, TKey3, TKey4>;
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1], TKey3 extends keyof TObject[TKey1][TKey2], TKey4 extends keyof TObject[TKey1][TKey2][TKey3]>(path: [TKey1, TKey2, TKey3, TKey4], object: TObject): TObject[TKey1][TKey2][TKey3][TKey4];
        <TObject extends object, TKey1 extends keyof TObject, TKey2 extends keyof TObject[TKey1], TKey3 extends keyof TObject[TKey1][TKey2], TKey4 extends keyof TObject[TKey1][TKey2][TKey3]>(path: [TKey1, TKey2, TKey3, TKey4], object: TObject | null | undefined): TObject[TKey1][TKey2][TKey3][TKey4] | undefined;
        (path: number): LodashPropertyOf9x1;
        <T>(path: lodash.__, object: lodash.NumericDictionary<T>): LodashPropertyOf9x2<T>;
        <T>(path: number, object: lodash.NumericDictionary<T>): T;
        <T>(path: lodash.__, object: lodash.NumericDictionary<T> | null | undefined): LodashPropertyOf10x2<T>;
        <T>(path: number, object: lodash.NumericDictionary<T> | null | undefined): T | undefined;
        (path: lodash.PropertyPath): LodashPropertyOf11x1;
        (path: lodash.__, object: null | undefined): LodashPropertyOf11x2;
        (path: lodash.PropertyPath, object: null | undefined): undefined;
        (path: lodash.__, object: any): LodashPropertyOf12x2;
        (path: lodash.PropertyPath, object: any): any;
    }
```

This was never found because there is no test with an object as first argument in the test file for `propertyOf`: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/35b81ce6b7865843f9c2e49861c470621134ce3b/types/lodash/lodash-tests.ts#L7060

As a workaround we can use `property` from `lodash/fp`

But it would still be nice to have this fixed?

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
It's not clear to me how `propertyOf` is intended to be called for fp. The documentation you linked shows that `propertyOf` has a single argument (arity of 1), but the [FP guide](https://github.com/lodash/lodash/wiki/FP-Guide) says that `propertyOf` has a fixed arity of 2, without explaining what those 2 arguments should be.

The generated types seem to suggest that `propertyOf` should be called with the property path first, then the object second. I'm not sure if that's correct, or if it's a bug in our FP type generator.

**Reply by aj-r** [reply]
I guess the generator must be wrong, since the current types would make `propertyOf` basically identical to `property` and `get`, but it should be the opposite.

EDIT: looks like the generator mistakenly marks propertyOf's arguments as "unchanged order" (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/lodash/scripts/generate-fp.ts#L183) which is probably why they are in the wrong order.

**Reply by aj-r** [reply]
We could fix the generator and regenerate the types, but there are other known bugs in the generator, so regeneration won't work correctly right now. So I propose:

1. Manually reverse the order of arguments in `fp.d.ts`
2. Either (a) update the generator script without running it, or (b) delete the generator script since it may be easier to just update the FP types by hand at this point.

**Reply by primeapple** [reply]
Thanks for your respose!

First about `propertyOf`: It was also unclear to me what the correct types are, so I tried it out. The first argument is always the `object` that contains the property at a given `path`. The second argument is the `path` to query the `object` with. If the second argument is not given, we return a `function`, if the second argument is given we return the entity that is in the `object` at the given `path`.

So the following would be valid:

```javascript
propertyOf({"a": 1})("a")); // -> 1
propertyOf({"a": 1}, "a")); // -> 1
```

`path` and `object` can be of different types, of course.

**Reply by primeapple** [reply]
I'm unsure what's the correct way to go on.  
I don't know how big the problems with the generator script are. Is it fixable or not worth the effort?

Maybe let's update the types by hand first, update the generator script and add a new discussion to maybe remove the generator script in the future?  We could collect all the issues with the script there and evaluate if it's worth fixing.

Could we also add test cases for `propertyOf`, tho?

**Reply by aj-r** [reply]
Yes that's probably the best approach (update by hand and fix the test cases).

The generator is fixable but probably not worth the effort. More issues keep popping up in the generator over time, so it's probably less effort to manually update the fp types with every change (and should result in better quality types).

**Reply by primeapple** [reply]
Thanks for your responses. I'll try to get a PR out to make this :)
